saleor/product/models.py

```diff
@@ -421,28 +449,89 @@ def get_absolute_url(self) -> str:
         return build_absolute_uri(url)
 
 
+class AttributeProduct(models.Model):
+    attribute = models.ForeignKey(
+        "Attribute", related_name="attributeproduct", on_delete=models.CASCADE
+    )
+    product_type = models.ForeignKey(
+        ProductType, related_name="attributeproduct", on_delete=models.CASCADE
+    )
+
+
+class AttributeVariant(models.Model):
+    attribute = models.ForeignKey(
+        "Attribute", related_name="attributevariant", on_delete=models.CASCADE
+    )
+    product_type = models.ForeignKey(
+        ProductType, related_name="attributevariant", on_delete=models.CASCADE
+    )
+
+
+class AttributeQuerySet(models.QuerySet):
+    def get_unassigned_attributes(self, product_type_pk: int):
+        return self.exclude(
+            Q(attributeproduct__product_type_id=product_type_pk)
+            | Q(attributevariant__product_type_id=product_type_pk)
+        )
+
+    def get_assigned_attributes(self, product_type_pk: int):
+        return self.filter(
+            Q(attributeproduct__product_type_id=product_type_pk)
+            | Q(attributevariant__product_type_id=product_type_pk)
+        )
+
+    @staticmethod
+    def user_has_access_to_all(user):
+        return user.is_active and user.has_perm("product.manage_products")
+
+    def get_public_attributes(self):
+        return self.filter(visible_in_storefront=True)
+
+    def get_visible_to_user(self, user):
+        if self.user_has_access_to_all(user):
+            return self.all()
+        return self.get_public_attributes()
```

Comment:
The mutation prefetching doesn't work. And is triggering a query for each attribute value (`prefetch_related("values")`).

This is also triggering a double queries (`prefetch` and `get_visible_to_user`).

We should have a `Prefetch` instead.